### PR TITLE
fix: collect-data should fail if CRs missing

### PR DESCRIPTION
**Pull request by johnbieren**

### CODE REVIEW

File: catalog/task/collect-data/0.1/collect-data.yaml [hunk 1]
```diff
@@ -38,7 +38,7 @@ spec:
       description: The first component's containerImage in the snapshot to use in the fbc pipelines
   steps:
     - name: collect-data
-      image: quay.io/hacbs-release/release-utils:f486f350df43974706fc484a13bfeefc2c8fa95a
+      image: quay.io/hacbs-release/release-utils:4d8649dbb2b626f5fe9f4ff83c1bc3be268fad31
```
Comment (mmalina) [inline catalog/task/collect-data/0.1/collect-data.yaml:41]: Why was this changed? There's nothing about it in the commit message. Is there a new functionality that is needed? If so, please link the corresponding PR. Or is it just to be on the latest image? That's fine, but please mention it in the commit message.

File: catalog/task/collect-data/0.1/collect-data.yaml [hunk 2]
```diff
@@ -53,32 +53,43 @@ spec:
       script: |
         #!/usr/bin/env sh
```
Comment (mmalina) [inline catalog/task/collect-data/0.1/collect-data.yaml:54]: Is there any reason why we don't want `set -e` here? Generally it's preferable to fail whenever something goes wrong if possible, because that makes it easier to debug stuff - rather than fail silently and carry on and then fail later on something and try to figure out why.

That way we wouldn't need the `if`s at all.

File: catalog/task/collect-data/0.1/collect-data.yaml [hunk 3]
```diff
@@ -53,32 +53,43 @@ spec:
...
 
-        if ! release_result=$(get-resource "release" "${RELEASE}" "{.spec.extraData}"); then
+        if ! get-resource "release" "${RELEASE}" > "$(workspaces.data.path)/release.json"; then
```
Comment (mmalina) [inline catalog/task/collect-data/0.1/collect-data.yaml:56]: So here we basically just switched the order - first we load the whole resources and then later we just get the `extraData` field for each. I think this is better. But effectively this shouldn't change the functionality, right? Because if the resource doesn't even exist, even previously it would fail on this, right? Or what am I missing?

File: catalog/task/collect-data/0.1/collect-data.yaml (hunk 2 above)
Comment (johnbieren) [reply catalog/task/collect-data/0.1/collect-data.yaml:54]: Because when you asked on the other PR about that, David said he didn't want to, so I assumed he still does not want it

File: catalog/task/collect-data/0.1/collect-data.yaml (hunk 3 above)
Comment (johnbieren) [reply catalog/task/collect-data/0.1/collect-data.yaml:56]: I just switched the order because I want to fail early if I am going to fail (if a CR is missing), instead of computing some jsons first then failing

File: catalog/task/collect-data/0.1/collect-data.yaml (hunk 1 above)
Comment (johnbieren) [reply catalog/task/collect-data/0.1/collect-data.yaml:41]: I updated the commit message. The new thing is https://github.com/redhat-appstudio/release-service-utils/pull/48

File: catalog/task/collect-data/0.1/collect-data.yaml (hunk 2 above)
Comment (mmalina) [reply catalog/task/collect-data/0.1/collect-data.yaml:54]: I do vaguely recall that, but I can't remember what the reasoning was. But ok.

File: catalog/task/collect-data/0.1/collect-data.yaml (hunk 3 above)
Comment (mmalina) [reply catalog/task/collect-data/0.1/collect-data.yaml:56]: OK, sure, but you agree it would fail anyway? The commit title suggests that only after this change it will now fail if a CR is missing.

File: catalog/task/collect-data/0.1/collect-data.yaml (hunk 1 above)
Comment (mmalina) [reply catalog/task/collect-data/0.1/collect-data.yaml:41]: Cool, thanks.

File: catalog/task/collect-data/0.1/collect-data.yaml (hunk 3 above)
Comment (johnbieren) [reply catalog/task/collect-data/0.1/collect-data.yaml:56]: It would fail anyway without the reordering, yes. It will not fail without this PR because there is no `exit 1` after checking the return code of the `get-resource > workspace` lines

File: catalog/task/collect-data/0.1/collect-data.yaml (hunk 2 above)
Comment (johnbieren) [reply catalog/task/collect-data/0.1/collect-data.yaml:54]: https://github.com/redhat-appstudio/release-service-bundles/pull/135#discussion_r1258383509
I think the `set -e` would be better, but he wrote it originally and seems to have a reason for it so I was trying to stick with not adding it

File: catalog/task/collect-data/0.1/collect-data.yaml (hunk 3 above)
Comment (mmalina) [reply catalog/task/collect-data/0.1/collect-data.yaml:56]: >  It will not fail without this PR because there is no exit 1 after checking the return code of the get-resource > workspace lines

But there is `exit 1` here at the top. For example, if the `release` resource is missing, this `if` will trigger and it will `exit 1`.
```sh
if ! release_result=$(get-resource "release" "${RELEASE}" "{.spec.extraData}");
```
Is that not the case?

Comment (johnbieren) [reply catalog/task/collect-data/0.1/collect-data.yaml:56]: Right. But what if the `snapshot` does not exist? We don't do anything with `extraData` from that. So I was consistent and changed to the check for all the retrievals to the workspace

File: catalog/task/collect-data/0.1/collect-data.yaml (hunk 2 above)
Comment (mmalina) [reply catalog/task/collect-data/0.1/collect-data.yaml:54]: Thanks for finding the comment. So I'm repeating myself when I say that I prefer `set -e` 

@davidmogar do you still insist in using `set +e`? Using `set -e` would make the script much simpler (no need for all the `if`s).

File: catalog/task/collect-data/0.1/collect-data.yaml (hunk 3 above)
Comment (mmalina) [reply catalog/task/collect-data/0.1/collect-data.yaml:56]: OK, fair enough. So can you please at least explain this in the commit message? That in most cases it would already fail, but for example if `snapshot` was missing, it would previously not fail?

Comment (johnbieren) [reply catalog/task/collect-data/0.1/collect-data.yaml:56]: Yeah, I will wait to see what happens with the `set -e` thing then adjust accordingly

File: catalog/task/collect-data/0.1/collect-data.yaml (hunk 2 above)
Comment (johnbieren) [reply catalog/task/collect-data/0.1/collect-data.yaml:54]: @davidmogar ping

File: catalog/task/collect-data/0.1/collect-data.yaml (hunk 3 above)
Comment (johnbieren) [reply catalog/task/collect-data/0.1/collect-data.yaml]: @mmalina fixed

File: catalog/task/collect-data/0.1/collect-data.yaml [hunk 4]
```diff
@@ -73,13 +61,20 @@ spec:
         get-resource "releaseplanadmission" "${RELEASE_PLAN_ADMISSION}" \
           > "$(workspaces.data.path)/release_plan_admission.json"
 
-        get-resource "releasestrategy" "${RELEASE_STRATEGY}" > "$(workspaces.data.path)/release_strategy.json"
```
Comment (mmalina) [inline catalog/task/collect-data/0.1/collect-data.yaml:76]: This looks like a mistake?

File: catalog/task/collect-data/0.1/collect-data.yaml (hunk 2 above)
Comment (mmalina) [reply catalog/task/collect-data/0.1/collect-data.yaml:54]: Cool, thanks.

File: catalog/task/collect-data/0.1/collect-data.yaml (hunk 4 above)
Comment (johnbieren) [reply catalog/task/collect-data/0.1/collect-data.yaml]: nice catch, fixed

File: catalog/task/collect-data/0.1/collect-data.yaml (hunk 3 above)
Comment (mmalina) [reply catalog/task/collect-data/0.1/collect-data.yaml]: cool, thanks

**Review by mmalina — APPROVED** [review]
